pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -48,13 +48,20 @@ import (
 const (
 	// Name of the plugin used in the plugin registry and configurations.
 	Name = "DefaultPreemption"
+	// Maximum number of candidates to shortlist when dry running preemption as a
+	// percentage of number of nodes.
+	maxCandidatesAsPercentageOfNodes = 10
```

Comment:
How about we use two parameters, percentage (`maxCandidatesAsPercentage`) and explicit value (`maxCandidates`). If set to higher than 0, `maxCandidates` takes precedence.  `minCandidates` should simply be hardcoded at 2 all the time.

I think this is a good case for the power of two choices: http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf

We need to think about how to make the selection random, perhaps we can randomize the starting offset of node search, but that might not be enough.

Reply:
One possible issue with that is in small-medium sized clusters with the default configuration. Consider a cluster with 50 nodes. Let's say a good default for `maxCandidatesAsPercentageOfNodes` is 10% (doesn't matter what it is as long as it's non-100%). If the cluster does not override `maxCandidatesAsPercentageOfNodes`, only 5 nodes will be enlisted. Is this sufficient? In such small clusters, the performance benefits between checking just 5 nodes will be outweighed by the better preemption we might get from checking all 50 nodes (which wouldn't be *that* slow).

Reply:
I am trying to allow using an absolute value without introducing a third parameter. I guess we can do that using the two parameters you have: if `maxCandidatesAsPercentage` is set to 0 `MinCandidates` becomes the maximum in absolute terms. Lets explicitly clarify that the number of nodes we look at is `max(nodeCount* MaxCandidatesAsPercentage, MinCandidates)`

Reply:
sorry, should have mentioned this earlier, it would have probably been easier if we split the PR into API change and core scheduler change (the api change requires api review), but it is fine since you did that already.

Reply:
@adtac Do you want to exercise using type `IntOrString` to represent an "either percentage or absolute number" semantic?

https://github.com/kubernetes/kubernetes/blob/f1012030f78ce73919ef426673b064809abc58dc/staging/src/k8s.io/api/core/v1/types.go#L4180

Reply:
re absolute maximum: I don't see the value of such an option immediately. While the 0% workaround would work, why would a cluster want to set an absolute value?

re the IntOrString type field -- I'm not a big fan of untyped fields, especially given YAML's reputation for being confusing :) I think it might be very confusing

>Lets explicitly clarify that the number of nodes we look at is max(nodeCount* MaxCandidatesAsPercentage, MinCandidates)

Done.